zeppelin-plugins/notebookrepo/mongo/src/main/java/org/apache/zeppelin/notebook/repo/MongoNotebookRepo.java

```diff
@@ -0,0 +1,390 @@
+   */
+  private void saveNotePath(String noteId, String noteName, String pId) {
+    Document filter = new Document(Fields.ID, noteId);
+    Document doc = new Document(Fields.ID, noteId)
+        .append(Fields.PID, pId)
+        .append(Fields.IS_DIR, false)
+        .append(Fields.NAME, noteName);
+
+    folders.replaceOne(filter, doc, new UpdateOptions().upsert(true));
+  }
+
+  private void saveNote(Note note) {
+    Document doc = noteToDocument(note);
+    notes.replaceOne(eq(Fields.ID, note.getId()), doc, new UpdateOptions().upsert(true));
+  }
+
+  @Override
+  public void move(String noteId, String notePath, String newNotePath
+      , AuthenticationInfo subject) throws IOException {
+    LOG.info("====move{}:{}:{}", noteId, notePath, newNotePath);
+    String[] pathArray = toPathArray(newNotePath, true);
+    String[] parentPathArray = Arrays.copyOfRange(pathArray, 0, pathArray.length - 1);
+    String noteName = pathArray[pathArray.length - 1];
+    String pId = completeFolder(parentPathArray);
+    moveNote(noteId, pId, noteName);
+  }
+
+  private void moveNote(String noteId, String parentId, String noteName) {
+    Document doc = new Document("$set"
+        , new Document(Fields.PID, parentId)
+        .append(Fields.NAME, noteName));
+
+    folders.updateOne(eq(Fields.ID, noteId), doc);
+    notes.updateOne(eq(Fields.ID, noteId), Updates.set(Fields.NAME, noteName));
+  }
+
+  @Override
+  public void move(String folderPath, String newFolderPath
+      , AuthenticationInfo subject) throws IOException {
+    LOG.info("====move folder {}:{}", folderPath, newFolderPath);
+    if (folderPath.equals(newFolderPath)) {
+      return;
+    }
+
+    String[] pathArray = toPathArray(folderPath, true);
+    String[] newPathArray = toPathArray(newFolderPath, true);
+    String[] newFolderParentArray = Arrays.copyOfRange(newPathArray, 0, newPathArray.length - 1);
+    String id = findFolder(pathArray);
+    String newPId = completeFolder(newFolderParentArray);
+    String newFolderName = newPathArray[newPathArray.length - 1];
+
+    Document doc = new Document("$set"
+        , new Document(Fields.ID, id)
+        .append(Fields.PID, newPId)
+        .append(Fields.IS_DIR, true)
+        .append(Fields.NAME, newFolderName));
+
+    folders.updateOne(eq(Fields.ID, id), doc);
+  }
+
+  @Override
+  public void remove(String noteId, String notePath
+      , AuthenticationInfo subject) throws IOException {
+    LOG.info("====remove {}:{}", noteId, notePath);
+    folders.deleteOne(eq(Fields.ID, noteId));
+    notes.deleteOne(eq(Fields.ID, noteId));
+
+    //clean empty folder
+    String[] pathArray = toPathArray(notePath, false);
+    for (int i = pathArray.length; i >= 0; i--) {
+      String[] current = Arrays.copyOfRange(pathArray, 0, i);
+      String folderId = findFolder(current);
+      boolean isEmpty = folders.count(eq(Fields.PID, folderId)) <= 0;
+      if (isEmpty) {
+        folders.deleteOne(eq(Fields.ID, folderId));
```

Comment:
this perhaps could have a race condition? if a new note is being added in the same folder

Reply:
It's quite a problem, I have two solutions:
- add read write lock
- upgrade to mongoDB 4.0, which have Multi-Document transaction.

which is better?